sql-serializer/src/main/java/io/cdap/pipeline/sql/serializer/bigquery/BigQuerySQLSerializer.java

```diff
@@ -0,0 +1,247 @@
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.pipeline.sql.serializer.bigquery;
+
+import io.cdap.pipeline.sql.api.Column;
+import io.cdap.pipeline.sql.api.Filter;
+import io.cdap.pipeline.sql.api.StructuredQuery;
+import io.cdap.pipeline.sql.api.Table;
+import io.cdap.pipeline.sql.api.constants.DateTimeConstant;
+import io.cdap.pipeline.sql.api.constants.IntegerConstant;
+import io.cdap.pipeline.sql.api.constants.StringConstant;
+import io.cdap.pipeline.sql.api.enums.OperandType;
+import io.cdap.pipeline.sql.api.enums.PredicateOperatorType;
+import io.cdap.pipeline.sql.api.enums.QueryableType;
+import io.cdap.pipeline.sql.api.interfaces.Constant;
+import io.cdap.pipeline.sql.api.interfaces.Operand;
+import io.cdap.pipeline.sql.api.interfaces.Queryable;
+import io.cdap.pipeline.sql.serializer.interfaces.SQLSerializer;
+
+import java.time.Instant;
+import java.time.ZoneOffset;
+import java.time.temporal.ChronoUnit;
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * Serializes a {@link StructuredQuery} into a {@link String} able to be executed upon Google BigQuery.
+ */
+public class BigQuerySQLSerializer implements SQLSerializer {
+  public String getSQL(StructuredQuery query) {
+    if (query == null) {
+      throw new IllegalArgumentException("Query cannot be null");
+    }
+    StringBuilder b = new StringBuilder();
+    return appendQuerySQL(b, query, false).toString();
+  }
+
+  public String getSQL(StructuredQuery query, Table temporaryTable) {
+    if (query == null) {
+      throw new IllegalArgumentException("Query cannot be null");
+    }
+    if (temporaryTable.hasAlias()) {
+      throw new IllegalArgumentException("Cannot have alias on temporary table");
+    }
+    if (temporaryTable.hasProject() || temporaryTable.hasDatabase()) {
+      throw new IllegalArgumentException("Temporary tables cannot be fully qualified");
+    }
+    StringBuilder b = new StringBuilder();
+    b.append("CREATE TEMPORARY TABLE ");
+    appendTableSQL(b, temporaryTable, false);
+    b.append(" AS ");
+    appendQuerySQL(b, query, false);
+    return b.toString();
+  }
+
+  private StringBuilder appendQuerySQL(StringBuilder b, StructuredQuery query, boolean isSubquery) {
+    // Keep track of which columns were defined in the current nesting level since they may also be used in filter
+    HashSet<Column> definedColumns = new HashSet<>();
+    if (isSubquery) {
+      b.append('(');
+    } else if (query.hasAlias()) {
+      throw new IllegalArgumentException("Cannot have an alias on the main outer query");
```

Comment:
this, for example, should only have to be checked once at the top, rather than once per subquery.

Reply:
Fixed